pkg/webhook/pod/tikv_creater.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package pod
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/pingcap/tidb-operator/pkg/apis/pingcap/v1alpha1"
+	"github.com/pingcap/tidb-operator/pkg/pdapi"
+	"github.com/pingcap/tidb-operator/pkg/webhook/util"
+	admission "k8s.io/api/admission/v1"
+	core "k8s.io/api/core/v1"
+	"k8s.io/klog"
+)
+
+const (
+	tikvNotBootstrapped = "\"TiKV cluster not bootstrapped, please start TiKV first\"\n"
+)
+
+func (pc *PodAdmissionControl) admitCreateTiKVPod(pod *core.Pod, tc *v1alpha1.TidbCluster, pdClient pdapi.PDClient) *admission.AdmissionResponse {
+
+	name := pod.Name
+	namespace := pod.Namespace
+
+	stores, err := pdClient.GetStores()
+	if err != nil {
+		if strings.HasSuffix(err.Error(), tikvNotBootstrapped) {
+			return util.ARSuccess()
+		}
+		klog.Infof("Failed to get stores during pod [%s/%s] creation, error: %v", namespace, name, err)
+		return util.ARFail(err)
+	}
+	evictLeaderSchedulers, err := pdClient.GetEvictLeaderSchedulers()
+	if err != nil {
+		if strings.HasSuffix(err.Error(), tikvNotBootstrapped) {
+			return util.ARSuccess()
+		}
+		klog.Infof("failed to create pod[%s/%s],%v", namespace, name, err)
+		return util.ARFail(err)
+	}
+
+	// if the pod which is going to be created already have a store and was in evictLeaderSchedulers,
+	// we should end this evict leader
+	for _, store := range stores.Stores {
```

Comment:
This too deeply nested block can be avoided by recording the schedulers in a map. And even better, when there is no evict leader schedulers, we don't have to loop over all the stores.

Reply:
updated